docs/proposals/auth.md

```diff
@@ -0,0 +1,150 @@
+  RC or namespace, while others should be allowed to push for any object in
+  the cluster.
+
+Proposed Solution
+-----------------
+
+Like in Kubernetes, authentication and authorization should be split into
+separate interfaces.  Each would be configured using a URL flag at the command
+line, similarly to how sinks and sources are currently configured.  Adding
+multiple authentication drivers would be supported (for example, to support
+authenticating from both client certificates and bearer tokens).
+
+### Authentication ###
+
+For authentication, the Kubernetes interface
+(`"k8s.io/kubernetes/pkg/auth/authenticator".Request`) can be reused so that
+any authentication plugins for Kubernetes can be easily used with Heapster.
+The interface is generic enough that it could also be used for specialized
+Heapster-specific interfaces.
+
+For example, if it was desired to use client certificate authentication,
+the command line flag might look like
+`--authn=clientca:/etc/heapster/ca.crt`.
+
+### Authorization ###
+
+For authorization, a system tailored to Heapster directly would be more useful:
+
+```go
+// This is basically just core.HistoricalKey from the Oldtimer PR,
+// renamed to be more generic
+type MetricKey struct {
+    // One of the core.MetricSetType* labels
+    ObjectType string
+
+    NodeName string
+    NamespaceName string
+    PodName string
+    ContainerName string
+    PodId string
+}
+
+type HeapsterQuerySource string
+var (
+    QuerySourceModel HeapsterQuerySource = "model"
+    QuerySourceHistorical HeapsterQuerySource = "historical"
+)
+
+type Authorizer interface {
+    // AuthorizeCreation authorizes creation of metrics with the given prefix
+    // (if any) and the given names at the given keys.
+    AuthorizeCreation(user kauthuser.Info, keys []MetricKey, prefix string, metricNames []string) error 
+
+    // AuthorizeQuery authorizes querying of metrics at the given keys with the
+    // given names from the given source.
+    AuthorizeQuery(user kauthuser.Info, source HeapsterQuerySource, keys []MetricKey, metricNames []string) error
+}
+```
+
+The `Authorizer` interface allows authorizing both queries against Heapster
+and requests to create new metrics entries.
+
+For queries, `AuthorizeQuery` takes user information, a source from which the
+information will be queried (the model or the historical source), and a set of
+keys and metric names.
+
+The authorizer may not assume a relationship between keys and metric names --
+it must assume all metric names given may be pushed for all metric keys (see
+below for the rationale for passing metric names).
+
+The authorizer will then take the given information, and return an error if
+the given user is not authorized to perform the given query.
+
+For creation of metrics, `AuthorizeCreation` takes much the same parameters
+as `AuthorizeQuery` (except for the `source` parameter), with the addition
+of the `prefix` parameter.  If the prefix parameter is given, when checking
```

Comment:
So, you have both kubernetes namespace, and "username prefix" as namespacing mechanisms.  Sorta confusing.